L1Trigger/L1THGCal/plugins/be_algorithms/HGCalSimCluster.cc

```diff
@@ -0,0 +1,256 @@
+// HGCal Trigger 
+#include "L1Trigger/L1THGCal/interface/HGCalTriggerBackendAlgorithmBase.h"
+//#include "L1Trigger/L1THGCal/interface/fe_codecs/HGCalBestChoiceCodec.h"
+#include "L1Trigger/L1THGCal/interface/fe_codecs/HGCalTriggerCellBestChoiceCodec.h"
+#include "DataFormats/ForwardDetId/interface/HGCTriggerDetId.h"
+
+// HGCalClusters and detId
+#include "DataFormats/L1THGCal/interface/HGCalCluster.h"
+#include "DataFormats/ForwardDetId/interface/HGCalDetId.h"
+
+// PF Cluster definition
+#include "DataFormats/ParticleFlowReco/interface/PFClusterFwd.h"
+#include "SimDataFormats/CaloAnalysis/interface/SimCluster.h"
+
+
+// Consumes
+#include "FWCore/Framework/interface/ConsumesCollector.h"
+#include "FWCore/Framework/interface/EDConsumerBase.h"
+
+//
+#include "DataFormats/Math/interface/LorentzVector.h"
+
+// Print out something before crashing or throwing exceptions
+#include <iostream>
+
+
+/* Original Author: Andrea Carlo Marini
+ * Original Dat: 23 Aug 2016
+ *
+ * This backend algorithm is supposed to use the sim cluster information to handle an
+ * optimal clustering algorithm, benchmark the performances of the enconding ...
+ */
+
+#define DEBUG
+
+
+
+namespace HGCalTriggerBackend{
+
+    template<typename FECODEC, typename DATA>
+        class HGCalTriggerSimCluster : public Algorithm<FECODEC>
+    {
+        private:
+            std::unique_ptr<l1t::HGCalClusterBxCollection> cluster_product;
+            // token
+            edm::EDGetTokenT< std::vector<SimCluster> > sim_token;
+            // handle
+            edm::Handle< std::vector<SimCluster> > sim_handle;
+            // add to cluster
+            void addToCluster(std::unordered_map<uint64_t,std::pair<int,l1t::HGCalCluster> >& cluster_container, uint64_t pid,int pdgid,float  energy,float  eta, float phi)
+            {
+            auto iterator = cluster_container.find (pid);
+            if (iterator == cluster_container.end())
+                {
+                    //create an empty cluster
+                    cluster_container[pid] = std::pair<int,l1t::HGCalCluster>(0,l1t::HGCalCluster());
+                    iterator = cluster_container.find (pid);
+                    iterator -> second . second . setPdgId(pdgid);
+                }
+            // p4 += p4' 
+            math::PtEtaPhiMLorentzVectorD p4;
```

Comment:
I don't think there is a += operator for candidates. But something like that should work:
```c++
pp4 += iterator -> second . second .p4();
iterator -> second . second . setP4(pp4);
```
This way you don't need to copy into p4